text/0000-buildpack-uris.md

```diff
@@ -0,0 +1,128 @@
+# Meta
+[meta]: #meta
+- Name: Buildpack URIs
+- Start Date: 02/07/2020
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: "N/A"
+
+# Summary
+[summary]: #summary
+
+A set of predefined URIs for referring to buildpacks from various sources (protocols). 
+
+_Please note that the scheme denotes protocol and not type._
+
+# Motivation
+[motivation]: #motivation
+
+Given that we support a multitude of sources for buildpacks, a certain level of ambiguity has naturally presented
+itself. We currently _try_ to resolve the various sources through a complex set of rules yet the ambiguity to users
+still exists. In config files, this ambiguity is at times resolved by introducing a separate key name to signify
+the source (protocol).
+
+The aim of this RFC is to propose a set of URI schemes that could be used in both CLI arguments and config files
+while solving the ambiguity problem.
+
+A few references to further express the problem:
+
+- [A poll](https://buildpacks.slack.com/archives/C94UJCNV6/p1580920669081100) to understand what users expect when
+providing buildpacks and there is overlap of sources.
+- [Issue](https://github.com/buildpacks/pack/issues/489) to provide additional clarity in logs due to ambiguity. 
+- [A conversation](https://buildpacks.slack.com/archives/CJ6B92ZSB/p1578502454017000) when trying to determine how to
+provide reference to all buildpacks in builder.
+- [Issue](https://github.com/buildpacks/pack/issues/376) where an `image` key was added to differentiate between
+directory and image. 
+ 
+# What it is
+[what-it-is]: #what-it-is
+
+### Scheme
+
+| Short Name | Format | Examples |
```

Comment:
There appears to be an inconsistency here in that only one of the URI schemes refers to a "type" of the thing being referenced.  For example, what if I want to refer to a builder at a URL or a buildpackage in the registry?  I think there are two ways of handling this.  Either the URI **never** has a reference to the type and the type is inferred contextually (`--buildpackage cnb://cloudfoundry/openjdk:1.2.0 --builder https://localhost/builder.tgz`) or the URI **always** contains the type to reduce ambiguity (`--buildpackage cnb:package:registry://cloudfoundry/openjdk:1.2.0 --builder cnb:builder:https://localhost/builder.tgz`).

I prefer the latter due to its extensibility and inability to be ambiguous:

```
cnb:[builder|package|dependency|...]:<any valid URI>
```
```
cnb:package:file:///./my/buildpack.tgz
cnb:builder:file:///home/user/my/builder.tgz
```
```
cnb:package:http://example.com/my/buildpack.tgz
cnb:builder:https://example.com/my/builder.tgz
```
```
cnb:package:docker:///mybuildpack:latest
cnb:builder:docker://gcr.io/mybuilder:latest
```
```
cnb:package:registry:///mybuildpack:latest
cnb:builder:registry://gcr.io/mybuilder:latest
```

Reply:
I tried to make it explicit in the summary but maybe haven't justified the why.

> Please note that the scheme denotes protocol and not type.

The "type" IMO should be handled via other methods and scheme should only detail protocol. Things like [media-type](https://github.com/opencontainers/image-spec/blob/master/media-types.md) or in our simplistic case for OCI images labels could be used.

Reply:
Then `cnb:builder://` has type information and shouldn't.

Reply:
It's not really a type. It's the "protocol" of how to search for buildpacks within a builder. That's at least my interpretation/justification.

Reply:
The RFC has been update to incorporate URNs to reference buildpacks in certain scenarios and additional considerations for registry resolved buildpacks.